File: TASC-Intern-Batch9/pharmacy-backend-microservice/order-service/src/main/java/com/pharmacy_backend/order_service/service/VoucherService.java
package com.pharmacy_backend.order_service.service;

import com.pharmacy_backend.common.dto.response.ApiResponse;
import com.pharmacy_backend.common.dto.response.PageResponse;
import com.pharmacy_backend.order_service.dto.request.UserVoucherRequest;
import com.pharmacy_backend.order_service.dto.request.VoucherRequest;
import com.pharmacy_backend.order_service.dto.response.VoucherResponse;

import java.util.List;

public interface VoucherService {
    ApiResponse<PageResponse<List<VoucherResponse>>> getVouchers(int pageIndex,
                                                                 int pageSize,
                                                                 String type);
    ApiResponse<VoucherResponse> getVoucherById(Long id);
    ApiResponse<Void> createVoucher(VoucherRequest request);
    ApiResponse<Void> updateVoucher(Long id,  VoucherRequest request);
    ApiResponse<Void> deleteVoucher(Long id);
    ApiResponse<Void> claimVoucher(UserVoucherRequest request);
    void changeVoucherStatus(Long voucherId, String status);
    ApiResponse<PageResponse<List<VoucherResponse>>> getUserVouchers(int pageIndex,
                                                                 int pageSize, String type);
}

File: pharmacy-v1/fe/fe-client/src/api/voucherService.ts
import type { ApiResponse, PageResponse } from '../types';
import type { Voucher, UserVoucherRequest } from '../types/voucher.types';
import axiosClient from './axiosClient';

const voucherService = {
  getVouchers: (pageIndex: number = 1, pageSize: number = 10, type?: string, status: string = 'ACTIVE') => {
    const params: any = { pageIndex, pageSize, status };
    if (type) params.type = type;
    return axiosClient.get<ApiResponse<PageResponse<Voucher[]>>>('/vouchers', { params });
  },

  getUserVouchers: (pageIndex: number = 1, pageSize: number = 10, type?: string, status?: string) => {
    const params: any = { pageIndex, pageSize };
    if (type) params.type = type;
    if (status) params.status = status;
    console.log("log")
    return axiosClient.get<ApiResponse<PageResponse<Voucher[]>>>('/vouchers/user/me', { params });
  },

  claimVoucher: (data: UserVoucherRequest) => {
    return axiosClient.post<ApiResponse<void>>('/vouchers/claim', data);
  },

  getVoucherById: (id: number) => {
    return axiosClient.get<ApiResponse<Voucher>>(`/vouchers/${id}`);
  }
};

export default voucherService;